### PR TITLE
chore: split test to upgrade and install parts

### PR DESCRIPTION
**Pull request by alejandroEsc**
**What type of PR is this?**

CHORE

**What this PR does/ why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
Helps testing in a more realistic way as we do not support deleting stateful addons and downgrading. 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Checklist**

* [ ] The commit message explains the changes and why are needed.
* [ ] The code builds and passes lint/style checks locally.
* [ ] The relevant subset of integration tests pass locally.
* [ ] The core changes are covered by tests.
* [ ] The documentation is updated where needed.

### CODE REVIEW

**Comment by s12chung** [comment]
```
 harness.go:73: ERROR: could not forward port to prometheus pod: pod prometheus-prometheus-kubeaddons-prom-prometheus-0 is not running, it's in phase Pending
 ```

**Comment by gracedo** [comment]
> ```
>  harness.go:73: ERROR: could not forward port to prometheus pod: pod prometheus-prometheus-kubeaddons-prom-prometheus-0 is not running, it's in phase Pending
> ```

This is an error we are also seeing in #778 , hoping increasing the sleep will fix that one

**Comment by alejandroEsc** [comment]
Seems to be an issue deploying traefik in one of the tests

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:496]: if no upgrade test needed, we probably don't need to actually do anything more for the addon right? this would essentially rerun `testGroupDeployment` which is just deploying the latest addon version

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:520]: same comment here: if no upgrade test needed, we probably don't need to actually do anything more for the addon right? this would essentially rerun `testGroupDeployment` which is just deploying the latest addon version

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:541]: I think at this point we should check if there are any addons that actually need to be upgrade-tested, if there are none then we should be able to skip running any of the steps here

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:270]: We shouldn't need to deploy the oldAddon - separately deploying an oldAddon that will be upgrade-tested isn't necessary since the `UpgradeAddon` plan actually [deploys the older version prior to upgrading](https://github.com/mesosphere/kubeaddons/blob/master/test/utils/addons.go#L289-L292), so in this case if we were to deploy the oldAddon first, the `UpgradeAddon` plan would delete the addon deployed (even if it's the old version) then redeploy the old version

**Review by gracedo — COMMENTED** [review]
Looks like finally tests passed (no upgrades), hooray!! For the upgrades test, which won't be triggered in this PR, I think overall we should be able to avoid running through two loops of addons and instead rework `checkIfUpgradeIsNeeded` to include all the logic. I hope my comments made sense, as I'm sure we all are aware, this test code is a bit hard to understand :) happy to go over this over zoom tomorrow if something doesn't make sense! Very possible I also may have missed something between this test code and the kubeaddons test utils code 😅

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:599]: That is something i tried to do, however, the issue is that the cluster reference is passed inito each upgrade object, and that reference is created on creation of the cluster. I could not seperate these concerns without actually changing the harness code.

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:217]: my comment above, this boolean for now allows us to get what addons to deploy first and (which some will get upgraded afte) (so a bit of a efficiency boos theer) and keeps a cluster from being unnecessarily deployed.

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:242]: Yes you are right!

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:263]: agreed

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:270]: I was missing this, thank you! We should only deploy those addons that are not upgraded, we do have to test the group after all.

**Review by gracedo — APPROVED** [review]
LGTM! I wonder if we want to run a dummy upgrade test on this PR before merging just to make sure the upgrade tests work? Maybe just bumping the revision on an addon, seeing that tests pass then reverting the change before merging?

**Comment by alejandroEsc** [comment]
> LGTM! I wonder if we want to run a dummy upgrade test on this PR before merging just to make sure the upgrade tests work? Maybe just bumping the revision on an addon, seeing that tests pass then reverting the change before merging?

That sounds good, let me branch from this and create a new pr but with a change.

File: test/addons_test.go
Comment (s12chung) [inline test/addons_test.go:329]: prefer calling this `addons`, as it's making me think of the old `addonDeployments`, but it's a nit

File: test/addons_test.go
Comment (s12chung) [inline test/addons_test.go:264]: i wonder why we need to do this comparison, shouldn't the controller handle it....

**Review by s12chung — APPROVED** [review]
overall, looks fine to me.

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:329]: hmmm, yeah i see what you mean, we can still make that change.

File: test/addons_test.go
Comment (alejandroEsc) [reply test/addons_test.go:264]: It is for the sake of creating addonUpgrades objects, which has little to do how kubernetes handles resources.

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:266]: oops, sorry I meant the `else` wasn't needed to wrap the log but let's keep the log in! (https://github.com/mesosphere/kubernetes-base-addons/pull/809/commits/5da80d493f3257b938d403c8ad42bff58af846b7#diff-6b1d50a89d4ac455215d0f0969ed4e5507e3a91528e4ef6a20e37d6d515781b2L263-L264)
```suggestion
		t.Logf("found old version of addon %s %s (revision %s) and new version %s (revision %s)", newAddon.GetName(), oldRev, oldVersion, newVersion, newRev)
		doUpgrade = true
```

File: test/addons_test.go
Comment (gracedo) [inline test/addons_test.go:610]: Can't comment on line 612 but below with `addonDefaults` plan (terribly named 🙄 maybe we should rename it to `waitForAddons` or something clearer lol) - This is waiting for all the addons including addons that haven't been deployed yet (ones that will be upgrade-tested, as they are deployed in the Upgrade plan) which is why we're seeing the tests fail in #817 . I think we should be passing in `deployments...` instead. `addonDefaults` is defined at [line 543](https://github.com/mesosphere/kubernetes-base-addons/pull/809/files#diff-6b1d50a89d4ac455215d0f0969ed4e5507e3a91528e4ef6a20e37d6d515781b2R543) (have to expand the diff to see)

**Comment by alejandroEsc** [comment]
closing in favor of PR #817

**Pull request closed without merging**
